draft-ietf-mls-protocol.md

```diff
@@ -914,6 +905,59 @@ np[0] -> 1       5[C]
 After performing these operations, the generator of the UpdatePath MUST
 delete the leaf_secret.
 
+## Adding and Removing Leaves
+
+In addition to the path-based updates to the tree described above, it is also
+necessary to add and remove leaves of the tree in order to reflect changes to the
+membership of the group (see {{add}} and {{remove}}).  Leaves are always added and removed at the
+right edge of the tree: Either a new rightmost leaf is added, or the rightmost
+leaf is removed.  Nodes' parent/child node relationships are then updated to
+maintain the tree's left-balanced structure.  These operations are also known as
+_extending_ and _truncating_ the tree.
+
+To add a new leaf: Add leaf L as the new rightmost leaf of the tree.  Add
+a blank parent node P whose right child is L.  P is attached to the
+tree as the right child of the only appropriate node to make the updated tree
+left-balanced (or set it as a new root).  The former right child of the P's
+parent becomes P's left child (or the old root becomes the P's left child if
+P is the new root).
+
+~~~~~
+                   _ <-- new parent              _
+                 __|_                          __|__
+                /    \                        /     \
+  X    ===>    X     |               ===>    X       _ <-- new parent
+ / \          / \    |                      / \     / \
+A   B        A   B   C <-- new leaf        A   B   C   D <-- new leaf
+~~~~~
+
+To remove the rightmost leaf: Remove the rightmost leaf node L and its parent
+node P.  If P was the root of the tree, P's left child
+is now the root of the tree.  Otherwise, set the right child of P's parent
+to be the P's left child.
+
+~~~~~
+      Y                                    Y
+    __|__                                __|_
+   /     \                              /    \
+  X       Z <-- remove parent  ===>    X     | <-- reassign child
+ / \     / \                          / \    |
+A   B   C   D <-- remove leaf        A   B   C
+
+
+      Y <-- remove parent
+    __|_
+   /    \
+  X     |                  ===>    X <-- reassign root
+ / \    |                         / \
+A   B   C <-- remove leaf        A   B
+~~~~~
+
```

Comment:
Even if this only describe the structure modification, I think it is valuable to give hints to the reader that in practice only blank leaves will be removed. If here the reader thinks that we can truncate any leaf and only afterward the reader discovers that only blank leaves can be truncated, I think it can be confusing.
Also it might look like a violation of the tree invariant, since in the second tree, D knows Y's secret without being its descendant.

```suggestion

Note: in the rest of the protocol, the rightmost leaf will be removed only when it is blank.

```

Reply:
Fair enough.  I think I will word this slightly differently, but I can grant that it makes sense to have a note here.